cEP-0002.md

```diff
@@ -0,0 +1,368 @@
+    * This would execute the marking of this bear as complete in the
+    dependency tracker.
+    * Once the marking is complete the tracker returns all bears which have no
+    dependencies remaining as pending, so that their tasks can be scheduled to
+    the process pool.
+
+Here is a prototype:
+
+```python
+from collections import defaultdict
+
+
+class Dependencies:
+    def __init__(self):
+        self.dependency_dict = defaultdict(list)
+        self.dependency_set = set()
+
+    def _add_dependency(self, bear_instance, dependency):
+        self.dependency_dict[bear_instance].append(dependency)
+        self.dependency_set.add(dependency)
+
+    def add_bear_dependencies(self, bear_instance):
+        """
+        Add all dependencies of a bear to the dependency_set() and dependency_dictionary.
+
+        :param bear_instance: The instance of the bear to whose dependencies
+                              to add.
+        """
+        for bear in bear_instance.BEAR_DEPS:
+            self._add_dependency(bear_instance, bear)
+
+    def resolve(self, bear_instance):
+        """
+        When a bear completes this method is called with the instance of that
+        bear. The method deletes this bear from the list of dependencies of
+        each bear in the dependency dictionary. It returns the bears which
+        have all of it's dependencies resolved.
+
+        :param bear_instance: The instance of the bear.
+        """
+        return_bears = []
+        if bear_instance.__class__ in self.dependency_set:
+            for bear in self.dependency_dict:
+                if bear_instance.__class__ in self.dependency_dict[bear]:
+                    self.dependency_dict[bear].remove(bear_instance.__class__)
+                    if not self.dependency_dict[bear]:
+                        return_bears.append(bear)
+        for bear in return_bears:
+            del self.dependency_dict[bear]
+        return return_bears
+```
+
+# A Prototype
+
+A rough look at how the new local bear class implementation will look:
+
+```python
+class LocalBear:
+    def __init__(self, section, result_queue, file_set):
+        """
+        Constructs a new bear.
+
+        :param section:      The section object where bear settings are
+                             contained.
+        :param result_queue: The queue object for messages. Can be ``None``.
+        :param file_set:     A set of FileProxy objects.
+        """
+        self.section = section
+        self.result_queue = result_queue
+        self.file_set = file_set
+
+        # May raise RuntimeError so bear doesn't get executed on invalid params
+        self.kwargs = get_kwargs_for_function(self.run, section)
+
+    def execute_task(self, task, **kwargs):
+        """
+        This method is responsible for getting results from the
+        analysis routines and populating the result queue.
+        """
+        return list(self.analyze(task, **self.kwargs))
+
+    def analyze(self,
```

Comment:
btw. we wanna totally be backwards compatible: I dislike the names GlobalBear and LocalBear so I'm for rewriting them such that they behave exactly like before but are using the new stuff behind the scenes. We're just writing some new classes like FileWideBear and ProjectWideBear which expose the new API like the fileproxy instead.